### PR TITLE
HDDS-9295. Avoid overriding finalize() in CodecBuffer.

### PR DESCRIPTION
**Pull request by szetszwo**
## What changes were proposed in this pull request?

CodecBuffer overrides the finalize() method for leak detection. It turns out causing a much longer GC time. (Thanks
@duongkame for finding it out.)

This JIRA is to change CodecBuffer not to override finalize(). In the test, we will create a subclass to override CodecBuffer for leak detection.

## What is the link to the Apache JIRA

HDDS-9295

## How was this patch tested?

Added a new test.

### CODE REVIEW

**Comment by umamaheswararao** [comment]
@szetszwo @duongkame Could you please update the JIRA description about the background of this issue?

**Comment by szetszwo** [comment]
@umamaheswararao , we do have the background information in the first two comments of HDDS-9295.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/db/CodecBuffer.java [hunk 1]
```diff
@@ -153,8 +188,11 @@ private void assertRefCnt(int expected) {
     Preconditions.assertSame(expected, buf.refCnt(), "refCnt");
   }
 
-  @Override
-  protected void finalize() throws Throwable {
+  /**
+   * Provide an implementation of the {@link #finalize()} method.
+   * For performance reason, this class does not override {@link #finalize()}.
+   */
+  void finalizeImpl() {
```
Comment (duongkame) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/db/CodecBuffer.java:195]: /nit can be named `detectLeaks`, no need to expose the details of finalize here.

**Review by duongkame — APPROVED** [review]
Thanks for the fix @szetszwo LGTM.
I think it's fine for now to just avoid the finalizer in Prod for now. However, at some point in time, we may need to establish a leak detection without hurting GC or prod performance, e.g. RocksObject. A finalizer alternative, like reference queue, can be considered (ref: https://www.oracle.com/technical-resources/articles/javase/finalization.html).

**Comment by szetszwo** [comment]
> ... A finalizer alternative, like reference queue, can be considered (ref: https://www.oracle.com/technical-resources/articles/javase/finalization.html).

ReferenceQueue sounds like a very good idea.

**Pull request merged**